src/proposer/proposer.h

```diff
@@ -38,6 +42,12 @@ class Proposer {
 
   virtual bool IsStarted() = 0;
 
+  virtual bool GenerateBlock(CWallet *wallet,
+                             const CBlockIndex &tip,
+                             const staking::CoinSet &coins,
+                             const boost::optional<CScript> &coinbase_script,
+                             std::shared_ptr<const CBlock> &block_out) = 0;
```

Comment:
I looked on implementation and usage of this function, and it seems we can return shared_ptr<const CBlock> right from the function (nullptr will mean `false` result).